### PR TITLE
Add tas-install role

### PR DESCRIPTION
**Pull request by mrymsza**
* deploy Prometheus
* deploy Prometheus Adapter
* deploy Prometheus Node Exporter
* deploy and configure Telemetry Aware Scheduler

### CODE REVIEW

File: roles/tas-install/charts/prometheus_custom_metrics_helm_chart/templates/custom-metrics-apiserver-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      serviceAccountName: custom-metrics-apiserver
+      containers:
+      - name: custom-metrics-apiserver
+        image: directxman12/k8s-prometheus-adapter-amd64
+        args:
+        - --secure-port=6443
+        - --tls-cert-file=/var/run/serving-cert/tls.crt
+        - --tls-private-key-file=/var/run/serving-cert/tls.key
+        - --logtostderr=true
+        - --prometheus-url=http://prom-service.monitoring.svc:9090/
```
Comment (przemeklal) [inline roles/tas-install/charts/prometheus_custom_metrics_helm_chart/templates/custom-metrics-apiserver-deployment.yaml]: please don't hardcode service names and namespaces names

File: roles/tas-install/charts/prometheus_helm_chart/values.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+configMapNamespace: monitoring
+
+deploymentName: prometheus-deployment
+deploymentNamespace: monitoring
+deploymentAppName: prometheus-server
+containerName:  prometheus
+containerImage: prom/prometheus:v2.2.1
+
+serviceName1: prometheus-service
+serviceName2: prom-service
```
Comment (przemeklal) [inline roles/tas-install/charts/prometheus_helm_chart/values.yaml]: what's the story behind these values? why do we need two?

File: roles/golang-install/defaults/main.yml [hunk 3]
```diff
@@ -1,3 +1,3 @@
 ---
-golang_version: 1.11.9
+golang_version: 1.12.16
```
Comment (przemeklal) [inline roles/golang-install/defaults/main.yml:2]: can this change be moved into a separate commit?

File: roles/cluster-defaults/defaults/main.yml [hunk 4]
```diff
@@ -13,6 +13,8 @@ no_proxy: >-
   {{ hostvars[item]['access_ip'] | d(hostvars[item]['ip'] | d(fallback_ips[item])) }},{{ item }},{{ item }}.{{ dns_domain }},
   {%- endfor -%}
   {%- if additional_no_proxy is defined -%}{{ additional_no_proxy }},{%- endif -%}
+  {%- if kube_service_addresses is defined -%}{{ kube_service_addresses }},{%- endif -%}
+  {%- if kube_pods_subnet is defined -%}{{ kube_pods_subnet }},{%- endif -%}
```
Comment (przemeklal) [inline roles/cluster-defaults/defaults/main.yml]: can this change be moved into a separate commit?

File: roles/tas-install/charts/prometheus_custom_metrics_helm_chart/templates/custom-metrics-apiserver-deployment.yaml (hunk 1 above)
Comment (przemeklal) [inline roles/tas-install/charts/prometheus_custom_metrics_helm_chart/templates/custom-metrics-apiserver-deployment.yaml]: it's not a template if everything is hardcoded - deployment names, namespaces, service names, image names, config map references, etc.

File: roles/tas-install/charts/prometheus_custom_metrics_helm_chart/templates/custom-metrics-apiserver-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  template:
+    metadata:
+      labels:
+        app: custom-metrics-apiserver
+      name: custom-metrics-apiserver
+    spec:
+      serviceAccountName: custom-metrics-apiserver
+      containers:
+      - name: custom-metrics-apiserver
+        image: directxman12/k8s-prometheus-adapter-amd64
```
Comment (przemeklal) [inline roles/tas-install/charts/prometheus_custom_metrics_helm_chart/templates/custom-metrics-apiserver-deployment.yaml]: please add image version tag, preferably make it configurable via helm values

File: roles/tas-install/charts/prometheus_custom_metrics_helm_chart/templates/custom-metrics-apiserver-resource-reader-cluster-role-binding.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: custom-metrics-resource-reader
```
Comment (przemeklal) [inline roles/tas-install/charts/prometheus_custom_metrics_helm_chart/templates/custom-metrics-apiserver-resource-reader-cluster-role-binding.yaml]: hardcoded

**Review by przemeklal — CHANGES_REQUESTED** [review]
Helm charts need to be fixed to allow configurability in terms of resource names, namespaces, image versions etc.

Also, there's a stable properly written chart for Prometheus adapter available, could we consider using it instead of writing our own one (which isn't going to be better) https://github.com/helm/charts/tree/master/stable/prometheus-adapter

This would be an external dependency, however that chart uses proper versioning, so we could reference a version that we're sure that it works with TAS.

File: roles/tas-install/tasks/prometheus_custom_metrics.yml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+- name: populate Prometheus Adapter rbac Helm chart values template and push to master node
+  template:
+    src: "prometheus-adapter-rbac-values.yml.j2"
+    dest: "/usr/src/charts/prometheus-adapter-rbac-values.yml"
+    force: yes
+  when:
+    - inventory_hostname == groups['kube-master'][0]
+
+- name: create rbac for node resource reader
+  command: kubectl create -f prometheus-adapter-rbac-values.yml
```
Comment (przemeklal) [inline roles/tas-install/tasks/prometheus_custom_metrics.yml]: this always fails on redeployment, please use `kubectl apply` instead or do delete and then create

File: roles/tas-install/tasks/tas.yml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+- name: set values for tas-extender
+  set_fact:
+    tas_extender_image: "{{ registry_local_address }}/tas-extender"
+    tas_extender_version: latest
+  when:
+    - inventory_hostname == groups['kube-master'][0]
+
+- name: generate extender-secret
+  command: kubectl create secret tls extender-secret --cert /etc/kubernetes/pki/ca.crt --key /etc/kubernetes/pki/ca.key
```
Comment (przemeklal) [inline roles/tas-install/tasks/tas.yml]: this always fails on redeployment, please do delete and then create as secret resources don't support `apply` command iirc

File: roles/tas-install/tasks/prometheus_node_exporter.yml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    force: yes
+  when:
+    - inventory_hostname == groups['kube-master'][0]
+
+- name: install Prometheus Node Exporter Helm chart
+  command: helm upgrade --install --namespace {{ prometheus_node_exporter_namespace }} {{ prometheus_node_exporter_name }} -f node-exporter-values.yml stable/prometheus-node-exporter
+  args:
+    chdir: "/usr/src/charts/"
+  when:
+    - inventory_hostname == groups['kube-master'][0]
```
Comment (przemeklal) [inline roles/tas-install/tasks/prometheus_node_exporter.yml]: this often fails on redeployment:
```
TASK [tas-install : install Prometheus Node Exporter Helm chart] **************************************************************************************************************************************************
skipping: [node1]
fatal: [master1]: FAILED! => {"changed": true, "cmd": ["helm", "upgrade", "--install", "--namespace", "kube-system", "node-exporter", "-f", "node-exporter-values.yml", "stable/prometheus-node-exporter"], "delta": "0:00:03.812645", "end": "2020-02-28 09:42:39.749309", "msg": "non-zero return code", "rc": 1, "start": "2020-02-28 09:42:35.936664", "stderr": "Error: UPGRADE FAILED: Could not get apiVersions from Kubernetes: unable to retrieve the complete list of server APIs: custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request", "stderr_lines": ["Error: UPGRADE FAILED: Could not get apiVersions from Kubernetes: unable to retrieve the complete list of server APIs: custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request"], "stdout": "UPGRADE FAILED\nError: Could not get apiVersions from Kubernetes: unable to retrieve the complete list of server APIs: custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request", "stdout_lines": ["UPGRADE FAILED", "Error: Could not get apiVersions from Kubernetes: unable to retrieve the complete list of server APIs: custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request"]}
```

purging all Prometheus Helm deployments before rerunning deployment seems to solve it, so maybe this is a way to fix it

File: roles/tas-install/tasks/tas.yml [hunk 10]
```diff
@@ -0,0 +1,95 @@
+---
+- name: build TAS
+  make:   
+    target: build
```
Comment (przemeklal) [inline roles/tas-install/tasks/tas.yml]: this seems to fail when git 1.x.x is on the host due to TAS golang deps

I think we need to add tasks in the host OS configuration layer to build and install git 2 from sources on RHEL/CentOS 7 if existing git version is lower than required by TAS

File: roles/tas-install/tasks/tas.yml [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    - inventory_hostname == groups['kube-master'][0]
+
+- name: install descheduler
+  make:
+    chdir: "{{ descheduler_dir }}"
+  when:
+    - inventory_hostname == groups['kube-master'][0]
+  
+- name: deploy TAS
+  command: kubectl apply -f deploy/
```
Comment (przemeklal) [inline roles/tas-install/tasks/tas.yml]: The biggest issue - this will always use local image defined in original TAS deployment yamls instead of the one pushed to cluster's Docker registry.

```
[root@master1 telemetry-aware-scheduling]# kubectl describe deployment telemetry-aware-scheduling
...
Name:                   telemetry-aware-scheduling
Namespace:              default
...
Pod Template:
...
  Containers:
   tascont:
    Image:      tas-controller
   ...
   tasext:
    Image:      tas-extender
   ...
```
Because of that it never deploys:
```
NAME                                          READY   STATUS             RESTARTS   AGE
telemetry-aware-scheduling-75596fd6b4-m9pcj   0/2     ImagePullBackOff   0          8m16s
```
Another side effect is that the namespace cannot be changed.

**Review by przemeklal — CHANGES_REQUESTED** [review]
There's still a couple of minor issues related mostly to the idempotency.

Beyond that deployment doesn't work because of hardcoded namespace and image tag in the deployment yamls hosted in TAS repo - we should turn them into jinja templates and put in this repository or, and that'd be the best solution, make a TAS Helm chart.

File: roles/tas-install/charts/intel-telemetry-aware-scheduler/Chart.yaml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+appVersion: "1.0"
+description: A Helm chart for Kubernetes
```
Comment (killianmuldoon) [inline roles/tas-install/charts/intel-telemetry-aware-scheduler/Chart.yaml]: Nit: Should this say something about TAS ? I'm not sure where this description is picked up

File: roles/tas-install/charts/intel-telemetry-aware-scheduler/templates/NOTES.txt [hunk 13]
```diff
@@ -0,0 +1,4 @@
+1. Get the application URL by running these commands:
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app.kubernetes.io/name={{ include "intel-telemetry-aware-scheduler.name" . }},app.kubernetes.io/instance={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
+  echo "Visit http://127.0.0.1:8080 to use your application"
+  kubectl port-forward $POD_NAME 8080:{{ .Values.service.port }}
```
Comment (killianmuldoon) [inline roles/tas-install/charts/intel-telemetry-aware-scheduler/templates/NOTES.txt]: Not sure I understand the utility of this section ?

File: roles/tas-install/charts/intel-telemetry-aware-scheduler/templates/tas-health-policy.yml [hunk 14]
```diff
@@ -0,0 +1,33 @@
+{{- if .Values.taspolicy.create }}
```
Comment (killianmuldoon) [inline roles/tas-install/charts/intel-telemetry-aware-scheduler/templates/tas-health-policy.yml]: Do we want a configurable policy created during creation? It doesn't seem useful given this is a dummy policy. Better to leave this out I think.

File: roles/tas-install/charts/intel-telemetry-aware-scheduler/values.yaml [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+  type: ClusterIP
+  port: 80
+
+tls:
+  enabled: false
+  # secretName: extender-secret
+  # key: <key>
+  # certificate: <secret>
+
+taspolicy:
```
Comment (killianmuldoon) [inline roles/tas-install/charts/intel-telemetry-aware-scheduler/values.yaml]: I think this can be taken out based on the above.

**Review by killianmuldoon — CHANGES_REQUESTED** [review]

**Comment by przemeklal** [comment]
@mrymsza I pushed a couple of fixes to increase the robustness of the role and chart and with these fixes included I managed to get it up and running on my setup.
Please squash commits with trivial fixes in that PR, add code for creation of the default TAS policy and we should be ready to merge if no further comments. Thanks.

**Review by przemeklal — APPROVED** [review]
